sonarqube/resource_sonarqube_qualitygate_usergroup_association.go

```diff
@@ -0,0 +1,176 @@
+}
+
+// GetQualityGateUsergroupAssociationProjects used in GetQualityGateUsergroupAssociation
+type GetQualityGateUsergroupAssociationProjects struct {
+	Login       string `json:"login,omitempty"`
+	Name        string `json:"name"`
+	Description string `json:"description,omitempty"`
+	Selected    bool   `json:"selected"`
+}
+
+// Returns the resource represented by this file.
+func resourceSonarqubeQualityGateUsergroupAssociation() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceSonarqubeQualityGateUsergroupAssociationCreate,
+		Read:   resourceSonarqubeQualityGateUsergroupAssociationRead,
+		Delete: resourceSonarqubeQualityGateUsergroupAssociationDelete,
+		// Importer: &schema.ResourceImporter{
+		// 	State: resourceSonarqubeQualityGateProjectAssociationImport,
+		// },
+
+		// Define the fields of this schema.
+		Schema: map[string]*schema.Schema{
+			"login_name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ExactlyOneOf: []string{"login_name", "group_name"},
+			},
+			"group_name": {
+				Type:         schema.TypeString,
+				Optional:     true,
+				ForceNew:     true,
+				ExactlyOneOf: []string{"login_name", "group_name"},
+			},
+			"gatename": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+		},
+	}
+}
+
+func resourceSonarqubeQualityGateUsergroupAssociationCreate(d *schema.ResourceData, m interface{}) error {
+	sonarQubeURL := m.(*ProviderConfiguration).sonarQubeURL
+	rawQuery := url.Values{
+		"gateName": []string{d.Get("gatename").(string)},
+	}
+
+	if target, ok := d.GetOk("login_name"); ok {
+		sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/qualitygates/add_user"
+		rawQuery.Add("login", target.(string))
+	} else if target, ok := d.GetOk("group_name"); ok {
+		sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/qualitygates/add_group"
+		rawQuery.Add("groupName", target.(string))
+	}
+	sonarQubeURL.RawQuery = rawQuery.Encode()
+
+	resp, err := httpRequestHelper(
+		m.(*ProviderConfiguration).httpClient,
+		"POST",
+		sonarQubeURL.String(),
+		http.StatusNoContent,
+		"resourceSonarqubeQualityGateUsergroupAssociationCreate",
+	)
+	defer resp.Body.Close()
+
+	if err != nil {
+		return fmt.Errorf("error creating Sonarqube quality gate association: %+v", err)
+	}
+
+	d.SetId(uuid.NewV4().String())
+	return resourceSonarqubeQualityGateUsergroupAssociationRead(d, m)
+}
+
+func resourceSonarqubeQualityGateUsergroupAssociationRead(d *schema.ResourceData, m interface{}) error {
+	sonarQubeURL := m.(*ProviderConfiguration).sonarQubeURL
+	sonarQubeURL.RawQuery = url.Values{
+		"gateName": []string{d.Get("gatename").(string)},
+		"selected": []string{"selected"},
+	}.Encode()
+
+	if _, ok := d.GetOk("login_name"); ok {
+		sonarQubeURL.Path = strings.TrimSuffix(sonarQubeURL.Path, "/") + "/api/qualitygates/search_users"
+	} else if _, ok := d.GetOk("group_name"); ok {
```

Comment:
here as well. the first if to check if we passed a login_name makes sense, but the second one is redundant since there isn't another option. If that makes sense